src/main/java/io/vertx/core/impl/launcher/commands/Watcher.java

```diff
@@ -0,0 +1,306 @@
+/*
+ *  Copyright (c) 2011-2013 The original author or authors
+ *  ------------------------------------------------------
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *       The Eclipse Public License is available at
+ *       http://www.eclipse.org/legal/epl-v10.html
+ *
+ *       The Apache License v2.0 is available at
+ *       http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.vertx.core.impl.launcher.commands;
+
+import com.sun.nio.file.SensitivityWatchEventModifier;
+import io.vertx.core.Handler;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static java.nio.file.StandardWatchEventKinds.*;
+
+/**
+ * A file alteration monitor based on a {@link WatchService} and watching files matching a set of includes patterns.
+ * These patterns are Ant patterns (can use {@literal **, * or ?}). This class takes 2 {@link Handler} as parameter
+ * and orchestrate the redeployment method when a matching file is modified (created, updated or deleted). Users have
+ * the possibility to execute a shell command during the redeployment. On a file change, the {@code undeploy}
+ * {@link Handler} is called, followed by the execution of the user command. Then the {@code deploy} {@link Handler}
+ * is invoked.
+ * <p/>
+ * The watcher watches all files from the current directory and sub-directories.
+ *
+ * @author Clement Escoffier <clement@apache.org>
+ */
+public class Watcher implements Runnable {
+
+  private final static Logger LOGGER = LoggerFactory.getLogger(Watcher.class);
+
+  private final List<String> includes;
+  private final File root;
+  private final Handler<Handler<Void>> deploy;
+  private final Handler<Handler<Void>> undeploy;
+  private final String cmd;
+
+  private WatchService service;
```

Comment:
Beware of WatchService - we used it in the old Redeployer originally but it didn't work properly on all OSes so I removed it and wrote our own watch code.

Maybe they've fixed those bugs - but I'm not sure...